binary-data-formats.qmd

```diff
@@ -0,0 +1,404 @@
+    The csv version of this data does not explicitly specify that the first column contains integers and the second column contains strings.
+    This produces ambiguity because the csv reader guesses that the data types are double and character respectively:
+
+    ```{r csv-ambiguity-failure}
+    csv_file <- tempfile(fileext = ".csv")
+    write_csv(original, csv_file)
+    read_csv(csv_file, show_col_types = FALSE)
+    ```
+
+    The `read_csv()` and `write_csv()` functions are designed to avoid this issue to an extent, but it is impossible to avoid it completely because the csv format is inherently ambiguous.
+
+-   Inaccuracy.
+    Plain text files store binary numeric data inaccurately due to precision limits.
+    We can verify this by writing 1000 random numbers to a csv file and then reloading the data from the file
+
+    ```{r csv-precision-failure}
+    #| message: false
+    csv_file <- tempfile(fileext = ".csv")
+    original <- tibble(old_x = rnorm(1000))
+    write_csv(original, csv_file)
+    reloaded <- read_csv(csv_file) |>
+      rename(new_x = old_x)
+    ```
+
+    Now let's compare the two, counting the number of times the original value is strictly identical to the reloaded one:
+
+    ```{r csv-precision-failure-2}
+    bind_cols(original, reloaded) |> 
+      mutate(same_value = old_x == new_x) |>
+      count(same_value)
+    ```
+
+    A substantial proportion of the reloaded values differ from the original ones.
+    These differences are very small in magnitude, and occur because the csv represents numeric values with limited precision.
+    Even so this can still be a source of problems in your analysis code.
+
+-   Poor performance.
+    Compared to other formats, csv files can be slow to read and write because the on-disk format is very different to how the data needs to be represented in-memory.
+    We'll see some examples of this later in the chapter.
+
+The Apache Parquet file format ([parquet.apache.org](https://parquet.apache.org/)) is designed to address these limitations.
+Like a csv file, a parquet file stores a single rectangular data set.
+Features of the parquet format that make it appealing for data scientists include:
+
+-   Parquet files store data in a binary format (not plain text), avoiding inaccuracies in encoding and making it faster to read and write than a csv file.
+-   Parquet files store metadata that uniquely identifies the data types stored in each column, which avoids the ambiguity of csv files.
+-   Parquet files are compressed by default, and are typically much smaller than the corresponding csv files would be.
+-   Parquet files internally break the data into distinct "chunks" each with associated metadata, making it possible to write intelligent file readers that can jump straight to a specific part of the data when the full data set is not needed.
+
+All these factors together make the parquet format appealing when working with larger data sets.
```

Comment:
Another point worth mentioning: Parquet is widely used by big data systems, so adopting it doesn't mean locking yourself or your collaborators into an R only workflow.

Reply:
Yeah, I'll piggyback on this comment to also say that it might be worthwhile to explicitly point out that though Apache Parquet is binary, it is an open standards-based format (as well as it being fairly widely adopted like Neal points out)

I will admit when I first encountered parquet I was personally hesitant because I thought "ugh some new proprietary binary format that I'll be stuck locked into". Even with the apache moniker in the intro, I'm not sure that would be a shibboleth of openness to everyone who is reading R4DS like it is to us.